File: src/main/java/codes/qdbp/serverplugin/Serverplugin.java
package codes.qdbp.serverplugin;

import codes.qdbp.serverplugin.listeners.*;
import codes.qdbp.serverplugin.commands.*;
import codes.qdbp.serverplugin.misc.FoodMap;
import codes.qdbp.serverplugin.recipes.LightRecipe;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.NamespacedKey;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.*;

public class Serverplugin extends JavaPlugin {

    private static final double currentPluginVersion = 2.3;

    public static ArrayList<UUID> afkPlayerList;

    public static HashMap<UUID, Integer> afkPlayerRunningTasksMap;

    public static HashMap<UUID, String> afkPlayerTimes;

    public static HashMap<UUID, Location> freecamPlayerLocation;


    @Override
    public void onEnable() {

        /*
        Add feature toggles to Config
         */
        getConfig().addDefault("Feature-Toggles.useAFK", true);
        getConfig().addDefault("Feature-Toggles.useBackpack", true);
        getConfig().addDefault("Feature-Toggles.useCraft", true);
        getConfig().addDefault("Feature-Toggles.useEnderchest", true);
        getConfig().addDefault("Feature-Toggles.useFreecam", true);
        getConfig().addDefault("Feature-Toggles.useUpgrade", true);
        getConfig().addDefault("Feature-Toggles.useSkipnight", true);
        getConfig().addDefault("Feature-Toggles.useSwitchworld", true);
        getConfig().addDefault("Feature-Toggles.useDeaths", true);
        getConfig().addDefault("Feature-Toggles.useEating", true);
        getConfig().addDefault("Feature-Toggles.useSleep", true);
        getConfig().addDefault("Feature-Toggles.useElytrachange", true);
        getConfig().addDefault("Feature-Toggles.useLightRecipe", true);
        getConfig().addDefault("Feature-Toggles.useDoubleOpenDoors", true);
        getConfig().options().copyDefaults(true);
        saveConfig();


        /*
        On - Off toggle
         */
        boolean useAFK = getConfig().getBoolean("Feature-Toggles.useAFK");
        boolean useBackpack = getConfig().getBoolean("Feature-Toggles.useBackpack");
        boolean useCraft = getConfig().getBoolean("Feature-Toggles.useCraft");
        boolean useEnderchest = getConfig().getBoolean("Feature-Toggles.useEnderchest");
        boolean useFreecam = getConfig().getBoolean("Feature-Toggles.useFreecam");
//        boolean useUpgrade = getConfig().getBoolean("Feature-Toggles.useUpgrade");
        boolean useSkipnight = getConfig().getBoolean("Feature-Toggles.useSkipnight");
        boolean useSwitchworld = getConfig().getBoolean("Feature-Toggles.useSwitchworld");
        boolean useDeaths = getConfig().getBoolean("Feature-Toggles.useDeaths");
        boolean useEating = getConfig().getBoolean("Feature-Toggles.useEating");
        boolean useSleep = getConfig().getBoolean("Feature-Toggles.useSleep");
        boolean useElytrachange = getConfig().getBoolean("Feature-Toggles.useElytrachange");
        boolean useLightRecipe = getConfig().getBoolean("Feature-Toggles.useLightRecipe");
        boolean useDoubleOpenDoors = getConfig().getBoolean("Feature-Toggles.useDoubleOpenDoors");

        freecamPlayerLocation = new HashMap<>();
        afkPlayerList = new ArrayList<>();
        afkPlayerRunningTasksMap = new HashMap<>();
        afkPlayerTimes = new HashMap<>();

        /*
        Commands
         */
        Objects.requireNonNull(getCommand("features")).setExecutor(new FeaturesCommand());
        Objects.requireNonNull(getCommand("info")).setExecutor(new InfoCommand());
        if (useAFK) Objects.requireNonNull(getCommand("afk")).setExecutor(new AFKCommand(this));
        if (useBackpack) Objects.requireNonNull(getCommand("backpack")).setExecutor(new BackpackCommand(this));
        if (useCraft) Objects.requireNonNull(getCommand("craft")).setExecutor(new CraftCommand());
        if (useEnderchest) Objects.requireNonNull(getCommand("enderchest")).setExecutor(new EnderChestCommand());
        if (useFreecam) Objects.requireNonNull(getCommand("freecam")).setExecutor(new FreecamCommand());
//        if (useUpgrade) Objects.requireNonNull(getCommand("upgrade")).setExecutor(new OpenUpgradeMenuCommand());
        if (useSkipnight) Objects.requireNonNull(getCommand("skipnight")).setExecutor(new SkipNightCommand(this));
        if (useSwitchworld) Objects.requireNonNull(getCommand("switchworld")).setExecutor(new SwitchWorldCommand());
        if (useDeaths) Objects.requireNonNull(getCommand("tode")).setExecutor(new TodeCommand(this));


        /*
        Listeners
         */
        if (useBackpack) getServer().getPluginManager().registerEvents(new InventoryCloseListener(this), this);
        if (useDeaths) getServer().getPluginManager().registerEvents(new InventoryListener(this), this);
        if (useDeaths) getServer().getPluginManager().registerEvents(new PlayerDeathListener(this), this);
        if (useElytrachange) getServer().getPluginManager().registerEvents(new PlayerInteractListener(), this);
        if (useEating) getServer().getPluginManager().registerEvents(new PlayerItemConsumeListener(FoodMap.getFoodMap()), this);
        getServer().getPluginManager().registerEvents(new PlayerJoinListener(this), this);
        if (useAFK) getServer().getPluginManager().registerEvents(new PlayerKickListener(this), this);
        if (useSleep) getServer().getPluginManager().registerEvents(new PlayerListener(this), this);
        if (useAFK) getServer().getPluginManager().registerEvents(new PlayerMoveListener(), this);
        if (useAFK || useFreecam) getServer().getPluginManager().registerEvents(new PlayerQuitListener(), this);
        if (useDoubleOpenDoors) getServer().getPluginManager().registerEvents(new PlayerDoorInteractListener(), this);


        /*
        Recipes
         */
        if (useLightRecipe) Bukkit.addRecipe(new LightRecipe(new NamespacedKey(this, "light")).getLightRecipe());
    }

    @Override
    public void onDisable() {
        freecamPlayerLocation.forEach((key, value) -> FreecamCommand.backportPlayer(Objects.requireNonNull(Bukkit.getPlayer(key))));
    }

    public static double getCurrentPluginVersion() {
        return currentPluginVersion;
    }
}

File: src/main/kotlin/codes/qdbp/serverplugin/Serverplugin.kt
package codes.qdbp.serverplugin

import codes.qdbp.serverplugin.commands.*
import codes.qdbp.serverplugin.listeners.*
import codes.qdbp.serverplugin.misc.*
import codes.qdbp.serverplugin.recipes.Recipes
import org.bukkit.Bukkit
import org.bukkit.NamespacedKey
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.util.logging.Level

class Serverplugin : JavaPlugin() {
    private val logger = Bukkit.getLogger()

    /**
     * Create DataStorages
     */
    val backpackStorage = DataStorage(File(dataFolder, "backpackStorage.yml"))
    val deathStorage = DataStorage(File(dataFolder, "deathStorage.yml"))
    val mapImageStorage = DataStorage(File(dataFolder, "mapImageStorage.yml"))

    /**
     * Namespaced Keys
     */
    val invisItemFrameNSK = NamespacedKey(this, "invisible")
    val efficiencyUpgradeNSK = NamespacedKey(this, "efficiency")

    override fun onEnable() {

        /**
         * Update Checker
         */
        if (!checkPluginUpToDate(this)) logger.log(Level.WARNING, "There is a newer plugin version available!")

        /**
         * Config Defaults
         */
        config.addDefault("Features.useAFK", true)
        config.addDefault("Features.useBackpack", true)
        config.addDefault("Features.useCraft", true)
        config.addDefault("Features.useEnderchest", true)
        config.addDefault("Features.useFreecam", true)
        config.addDefault("Features.useSkipNight", true)
        config.addDefault("Features.useSwitchWorld", true)
        config.addDefault("Features.useDeaths", true)
        config.addDefault("Features.useEating", true)
        config.addDefault("Features.useSleep", true)
        config.addDefault("Features.useLightRecipe", true)
        config.addDefault("Features.useDoubleOpenDoors", true)
        config.addDefault("Features.useCustomMapImages", true)
        config.addDefault("Features.useUpgrade", true)
        config.addDefault("Features.useInvisibleItemFrames", true)
        config.addDefault("Features.useChangeUpgrade", true)
        config.addDefault("Features.useFastLeavesDecay", true)
        config.options().copyDefaults(true)
        saveConfig()

        /**
         * Get Config Toggles
         */
        val useAFK = config.getBoolean("Features.useAFK")
        val useBackpack = config.getBoolean("Features.useBackpack")
        val useCraft = config.getBoolean("Features.useCraft")
        val useEnderchest = config.getBoolean("Features.useEnderchest")
        val useFreecam = config.getBoolean("Features.useFreecam")
        val useSkipNight = config.getBoolean("Features.useSkipNight")
        val useSwitchWorld = config.getBoolean("Features.useSwitchWorld")
        val useDeaths = config.getBoolean("Features.useDeaths")
        val useEating = config.getBoolean("Features.useEating")
        val useSleep = config.getBoolean("Features.useSleep")
        val useLightRecipe = config.getBoolean("Features.useLightRecipe")
        val useDoubleOpenDoors = config.getBoolean("Features.useDoubleOpenDoors")
        val useCustomMapImages = config.getBoolean("Features.useCustomMapImages")
        val useUpgrade = config.getBoolean("Features.useUpgrade")
        val useInvisibleItemFrames = config.getBoolean("Features.useInvisibleItemFrames")
        val useChangeUpgrade = config.getBoolean("Features.useChangeUpgrade")
        val useFastLeavesDecay = config.getBoolean("Features.useFastLeavesDecay")

        /**
         * Commands
         */
        getCommand("use")?.setExecutor(UseCmd(this))
        getCommand("features")?.setExecutor(FeaturesCmd())
        getCommand("info")?.setExecutor(InfoCmd())
        if (useAFK) getCommand("afk")?.setExecutor(AfkCmd(this))
        if (useBackpack) getCommand("backpack")?.setExecutor(BackpackCmd(this))
        if (useCraft) getCommand("craft")?.setExecutor(CraftCmd())
        if (useEnderchest) getCommand("enderchest")?.setExecutor(EnderchestCmd())
        if (useFreecam) getCommand("freecam")?.setExecutor(FreecamCmd())
        if (useSwitchWorld) getCommand("switchworld")?.setExecutor(SwitchWorldCmd())
        if (useDeaths) getCommand("deaths")?.setExecutor(DeathsCmd(this))
        if (useSkipNight) getCommand("skipnight")?.setExecutor(SkipNightCmd(this))
        if (useCustomMapImages) getCommand("mapimage")?.setExecutor(MapImageCmd(this))
        if (useUpgrade) getCommand("upgrade")?.setExecutor(UpgradeCmd(this))
        if (useChangeUpgrade) getCommand("changeUpgrade")?.setExecutor(EditUpgradesCmd(this))

        /**
         * Listeners
         */
        server.pluginManager.registerEvents(PlayerJoinListener(this), this)
        if (useBackpack) server.pluginManager.registerEvents(InventoryCloseListener(this), this)
        if (useDeaths) {
            server.pluginManager.registerEvents(InventoryListener(this), this)
            server.pluginManager.registerEvents(PlayerDeathListener(this), this)
        }
        if (useEating) server.pluginManager.registerEvents(PlayerItemConsumeListener(), this)
        if (useAFK) server.pluginManager.registerEvents(PlayerKickListener(this), this)
        if (useSleep) server.pluginManager.registerEvents(PlayerBedEnterListener(this), this)
        if (useAFK) server.pluginManager.registerEvents(PlayerMoveListener(), this)
        if (useAFK || useFreecam) server.pluginManager.registerEvents(PlayerQuitListener(this), this) //TODO dont know about this one
        if (useDoubleOpenDoors) server.pluginManager.registerEvents(PlayerDoorInteractListener(), this) //TODO fix sneaking
        if (useInvisibleItemFrames) server.pluginManager.registerEvents(PlayerItemFrameChange(this), this)
        if (useInvisibleItemFrames) server.pluginManager.registerEvents(HangingPlaceEvent(this), this)
        if (useInvisibleItemFrames) server.pluginManager.registerEvents(BreakThisListenern(this), this)
        if (useFastLeavesDecay) server.pluginManager.registerEvents(LeavesDecayListener(), this)

        /**
         * Recipes
         */
        if (useLightRecipe) Bukkit.addRecipe(Recipes(this).lightRecipe)
        if (useInvisibleItemFrames) Bukkit.addRecipe(Recipes(this).invisItemFrameRecipe)

        //Reload MapImages
        mapImageStorage.storage.getIntegerList("mapIDs").forEach {
            val map = Bukkit.getMap(it)
            while (map?.renderers?.iterator()?.hasNext() ?: return) {
                map.removeRenderer(map.renderers.iterator().next())
            }
            MapImage(this, map)
        }
    }

    override fun onDisable() {

        /**
         * Return freecamplayers
         */
        if (freecamPlayers.isNotEmpty()) {
            freecamPlayers.forEach {
                freecamPlayers[it.key]!!.player.endFreecam()
            }
        }
    }
}